matrix_sdk_crypto/src/machine.rs

```diff
@@ -930,10 +928,30 @@ impl OlmMachine {
             .get(&KeyAlgorithm::Ed25519)
             .ok_or(EventError::MissingSigningKey)?;
 
-        Ok((
-            serde_json::from_value::<EventJson<AnyToDeviceEvent>>(decrypted_json)?,
-            signing_key.to_owned(),
-        ))
+        let ev_type = decrypted_json
+            .get("type")
+            .cloned()
+            .ok_or_else(|| EventError::MissingField("type".to_string()))?;
+
+        let content = decrypted_json
+            .get("content")
+            .cloned()
+            .ok_or_else(|| EventError::MissingField("content".to_string()))?;
+
+        let sender = decrypted_json
+            .get("sender")
+            .cloned()
+            .ok_or_else(|| EventError::MissingField("sender".to_string()))?;
+
+        let device_event = json!({
+            "type": ev_type,
+            "content": content,
+            "sender": sender
+        });
+
+        let event = serde_json::from_value::<EventJson<AnyToDeviceEvent>>(device_event)?;
```

Comment:
The old `ToDevice` event had a custom `Deserialize` impl so picking apart the object is needed now.

Reply:
Aren't you here pulling out fields that are inside a json value and then constructing the json value with the same format and data? How do the two json values differ?

Reply:
The derive `Deserialize` will fail if it finds a key that isn't in the rust struct. This failed to deserialize because of it. It took me a while to realize that the custom one https://github.com/ruma/ruma-events/blob/0.21.2/src/to_device.rs#L148 was just picking appart and ignoring the rest.